synthesizer/process/src/finalize.rs

```diff
@@ -84,46 +86,30 @@ impl<N: Network> Process<N> {
-            let stack = self.get_stack(transition.program_id())?;
-            // Ensure the number of calls matches the number of transitions.
-            let number_of_calls = stack.get_number_of_calls(transition.function_name())?;
-            ensure!(
-                number_of_calls == execution.len(),
-                "The number of transitions in the execution is incorrect. Expected {number_of_calls}, but found {}",
-                execution.len()
-            );
-        }
+        // Retrieve the root transition (without popping it).
+        let transition = execution.peek()?;
+        // Retrieve the stack.
+        let stack = self.get_stack(transition.program_id())?;
+        // Ensure the number of calls matches the number of transitions.
+        let number_of_calls = stack.get_number_of_calls(transition.function_name())?;
+        ensure!(
+            number_of_calls == execution.len(),
+            "The number of transitions in the execution is incorrect. Expected {number_of_calls}, but found {}",
+            execution.len()
+        );
         lap!(timer, "Verify the number of transitions");
 
+        // Construct the call graph.
+        let call_graph = self.construct_call_graph(execution)?;
+
         atomic_batch_scope!(store, {
             // Initialize a list for finalize operations.
             let mut finalize_operations = Vec::new();
 
-            /* Finalize the execution. */
-
-            // TODO (howardwu): This is a temporary approach. We should create a "CallStack" and recurse through the stack.
-            //  Currently this loop assumes a linearly execution stack.
-            // Finalize each transition, starting from the last one.
-            for transition in execution.transitions() {
-                // Retrieve the program ID.
-                let program_id = transition.program_id();
-                // Retrieve the function name.
-                let function_name = transition.function_name();
-                // Retrieve the stack.
-                let stack = self.get_stack(program_id)?;
-                // Finalize the transition.
-                match finalize_transition(state, store, stack, transition) {
-                    // If the evaluation succeeds with an operation, add it to the list.
-                    Ok(operations) => finalize_operations.extend(operations),
-                    // If the evaluation fails, bail and return the error.
-                    Err(error) => bail!("'finalize' failed on '{program_id}/{function_name}' - {error}"),
-                }
-                lap!(timer, "Finalize transition for '{program_id}/{function_name}'");
-            }
+            // Finalize the root transition.
+            // Note that this will result in all the remaining transitions being finalized, since the number
+            // of calls matches the number of transitions.
+            finalize_operations.extend(finalize_transition(state, store, stack, transition, call_graph)?);
```

Comment:
this can just be assigned to `finalize_operations`, without the need for `extend`